Display version, our license, and vendor attributions in version command

Kourin1996 opened the pull request:
Fixes EDGE-376 and EDGE-391

**Pending to merge because this PR requires go v1.16. Please upgrade go version before merging this PR.**

# Description

This PR sets the proper version which is used in `version` command and adds `license` command to print our license and dependency attributions.

1. Print proper version in `version` command

Version can be embedded by `-ldflags` on build time
+ Release products: The tag name will be embeded as version.
+ Local build (`make build`): The latest version + current commit short hash will be embeded.

2. Print our license in `license` command

Display our license written in `LICENSE` file.

3. Print vendor attributions in `license` command 

Display vendor attributions defined in version/bsd_licenses.json for BSD-2/3 license packages

## Result

`version` command prints the following results:

```

[POLYGON EDGE VERSION]
v0.2.13

```

`license` command prints the following results:

```

[LICENSE]

                                 Apache License
                           Version 2.0, January 2004
                        http://www.apache.org/licenses/

   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION

   1. Definitions.

      "License" shall mean the terms and conditions for use, reproduction,
      and distribution as defined by Sections 1 through 9 of this document.

      "Licensor" shall mean the copyright owner or entity authorized by
      the copyright owner that is granting the License.

      "Legal Entity" shall mean the union of the acting entity and all
      other entities that control, are controlled by, or are under common
      control with that entity. For the purposes of this definition,
      "control" means (i) the power, direct or indirect, to cause the
      direction or management of such entity, whether by contract or
      otherwise, or (ii) ownership of fifty percent (50%) or more of the
      outstanding shares, or (iii) beneficial ownership of such entity.

      "You" (or "Your") shall mean an individual or Legal Entity
      exercising permissions granted by this License.

      "Source" form shall mean the preferred form for making modifications,
      including but not limited to software source code, documentation
      source, and configuration files.

      "Object" form shall mean any form resulting from mechanical
      transformation or translation of a Source form, including but
      not limited to compiled object code, generated documentation,
      and conversions to other media types.

      "Work" shall mean the work of authorship, whether in Source or
      Object form, made available under the License, as indicated by a
      copyright notice that is included in or attached to the work
      (an example is provided in the Appendix below).

      "Derivative Works" shall mean any work, whether in Source or Object
      form, that is based on (or derived from) the Work and for which the
      editorial revisions, annotations, elaborations, or other modifications
      represent, as a whole, an original work of authorship. For the purposes
      of this License, Derivative Works shall not include works that remain
      separable from, or merely link (or bind by name) to the interfaces of,
      the Work and Derivative Works thereof.

      "Contribution" shall mean any work of authorship, including
      the original version of the Work and any modifications or additions
      to that Work or Derivative Works thereof, that is intentionally
      submitted to Licensor for inclusion in the Work by the copyright owner
      or by an individual or Legal Entity authorized to submit on behalf of
      the copyright owner. For the purposes of this definition, "submitted"
      means any form of electronic, verbal, or written communication sent
      to the Licensor or its representatives, including but not limited to
      communication on electronic mailing lists, source code control systems,
      and issue tracking systems that are managed by, or on behalf of, the
      Licensor for the purpose of discussing and improving the Work, but
      excluding communication that is conspicuously marked or otherwise
      designated in writing by the copyright owner as "Not a Contribution."

      "Contributor" shall mean Licensor and any individual or Legal Entity
      on behalf of whom a Contribution has been received by Licensor and
      subsequently incorporated within the Work.

   2. Grant of Copyright License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      copyright license to reproduce, prepare Derivative Works of,
      publicly display, publicly perform, sublicense, and distribute the
      Work and such Derivative Works in Source or Object form.

   3. Grant of Patent License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      (except as stated in this section) patent license to make, have made,
      use, offer to sell, sell, import, and otherwise transfer the Work,
      where such license applies only to those patent claims licensable
      by such Contributor that are necessarily infringed by their
      Contribution(s) alone or by combination of their Contribution(s)
      with the Work to which such Contribution(s) was submitted. If You
      institute patent litigation against any entity (including a
      cross-claim or counterclaim in a lawsuit) alleging that the Work
      or a Contribution incorporated within the Work constitutes direct
      or contributory patent infringement, then any patent licenses
      granted to You under this License for that Work shall terminate
      as of the date such litigation is filed.

   4. Redistribution. You may reproduce and distribute copies of the
      Work or Derivative Works thereof in any medium, with or without
      modifications, and in Source or Object form, provided that You
      meet the following conditions:

      (a) You must give any other recipients of the Work or
          Derivative Works a copy of this License; and

      (b) You must cause any modified files to carry prominent notices
          stating that You changed the files; and

      (c) You must retain, in the Source form of any Derivative Works
          that You distribute, all copyright, patent, trademark, and
          attribution notices from the Source form of the Work,
          excluding those notices that do not pertain to any part of
          the Derivative Works; and

      (d) If the Work includes a "NOTICE" text file as part of its
          distribution, then any Derivative Works that You distribute must
          include a readable copy of the attribution notices contained
          within such NOTICE file, excluding those notices that do not
          pertain to any part of the Derivative Works, in at least one
          of the following places: within a NOTICE text file distributed
          as part of the Derivative Works; within the Source form or
          documentation, if provided along with the Derivative Works; or,
          within a display generated by the Derivative Works, if and
          wherever such third-party notices normally appear. The contents
          of the NOTICE file are for informational purposes only and
          do not modify the License. You may add Your own attribution
          notices within Derivative Works that You distribute, alongside
          or as an addendum to the NOTICE text from the Work, provided
          that such additional attribution notices cannot be construed
          as modifying the License.

      You may add Your own copyright statement to Your modifications and
      may provide additional or different license terms and conditions
      for use, reproduction, or distribution of Your modifications, or
      for any such Derivative Works as a whole, provided Your use,
      reproduction, and distribution of the Work otherwise complies with
      the conditions stated in this License.

   5. Submission of Contributions. Unless You explicitly state otherwise,
      any Contribution intentionally submitted for inclusion in the Work
      by You to the Licensor shall be under the terms and conditions of
      this License, without any additional terms or conditions.
      Notwithstanding the above, nothing herein shall supersede or modify
      the terms of any separate license agreement you may have executed
      with Licensor regarding such Contributions.

   6. Trademarks. This License does not grant permission to use the trade
      names, trademarks, service marks, or product names of the Licensor,
      except as required for reasonable and customary use in describing the
      origin of the Work and reproducing the content of the NOTICE file.

   7. Disclaimer of Warranty. Unless required by applicable law or
      agreed to in writing, Licensor provides the Work (and each
      Contributor provides its Contributions) on an "AS IS" BASIS,
      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
      implied, including, without limitation, any warranties or conditions
      of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
      PARTICULAR PURPOSE. You are solely responsible for determining the
      appropriateness of using or redistributing the Work and assume any
      risks associated with Your exercise of permissions under this License.

   8. Limitation of Liability. In no event and under no legal theory,
      whether in tort (including negligence), contract, or otherwise,
      unless required by applicable law (such as deliberate and grossly
      negligent acts) or agreed to in writing, shall any Contributor be
      liable to You for damages, including any direct, indirect, special,
      incidental, or consequential damages of any character arising as a
      result of this License or out of the use or inability to use the
      Work (including but not limited to damages for loss of goodwill,
      work stoppage, computer failure or malfunction, or any and all
      other commercial damages or losses), even if such Contributor
      has been advised of the possibility of such damages.

   9. Accepting Warranty or Additional Liability. While redistributing
      the Work or Derivative Works thereof, You may choose to offer,
      and charge a fee for, acceptance of support, warranty, indemnity,
      or other liability obligations and/or rights consistent with this
      License. However, in accepting such obligations, You may act only
      on Your own behalf and on Your sole responsibility, not on behalf
      of any other Contributor, and only if You agree to indemnify,
      defend, and hold each Contributor harmless for any liability
      incurred by, or claims asserted against, such Contributor by reason
      of your accepting any such warranty or additional liability.

   END OF TERMS AND CONDITIONS

   APPENDIX: How to apply the Apache License to your work.

      To apply the Apache License to your work, attach the following
      boilerplate notice, with the fields enclosed by brackets "[]"
      replaced with your own identifying information. (Don't include
      the brackets!)  The text should be enclosed in the appropriate
      comment syntax for the file format. We also recommend that a
      file or class name and description of purpose be included on the
      same "printed page" as the copyright notice for easier
      identification within third-party archives.

   Copyright [yyyy] [name of copyright owner]

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


[DEPENDENCY ATTRIBUTIONS]

   This product bundles golang.org/x/crypto v0.0.0-20220126234351-aa10faf2a1f8,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/golang.org/x/crypto/LICENSE.

   This product bundles golang.org/x/net v0.0.0-20220127074510-2fabfed7e28f,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/golang.org/x/net/LICENSE.

   This product bundles golang.org/x/sys v0.0.0-20220114195835-da31bd327af9,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/golang.org/x/sys/LICENSE.

   This product bundles golang.org/x/text v0.3.7,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/golang.org/x/text/LICENSE.

   This product bundles golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/golang.org/x/time/LICENSE.

   This product bundles golang.org/x/sync v0.0.0-20210220032951-036812b2e83c,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/golang.org/x/sync/LICENSE.

   This product bundles golang/protobuf v1.5.2,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/google.golang.org/protobuf/LICENSE.

   This product bundles square/go-jose.v2 v2.5.1,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/gopkg.in/square/go-jose.v2/json/LICENSE.

   This product bundles flynn/noise v1.0.0,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/flynn/noise/LICENSE.

   This product bundles gogo/protobuf v1.3.2,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/gogo/protobuf/LICENSE.

   This product bundles golang/protobuf v1.5.2,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/golang/protobuf/LICENSE.

   This product bundles golang/snappy v0.0.4,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/golang/snappy/LICENSE.

   This product bundles google/gopacket v1.1.19,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/google/gopacket/LICENSE.

   This product bundles google/uuid v1.3.0,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/google/uuid/LICENSE.

   This product bundles imdario/mergo v0.3.12,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/imdario/mergo/LICENSE.

   This product bundles klauspost/compress v1.14.2,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/klauspost/compress/LICENSE.

   This product bundles libp2p/go-netroute v0.2.0,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/libp2p/go-netroute/LICENSE.

   This product bundles libp2p/go-sockaddr v0.1.0,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/libp2p/go-sockaddr/LICENSE.

   This product bundles multiformats/go-base32 v0.0.4,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/multiformats/go-base32/LICENSE.

   This product bundles pierrec/lz4 v2.5.2,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/pierrec/lz4/LICENSE.

   This product bundles pmezard/go-difflib v1.0.0,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/pmezard/go-difflib/LICENSE.

   This product bundles spaolacci/murmur3 v1.1.0,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/spaolacci/murmur3/LICENSE.

   This product bundles umbracle/go-eth-bn256 v0.0.0-20190607160430-b36caf4e0f6b,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/umbracle/go-eth-bn256/LICENSE.

   This product bundles marten-seemann/qtls-go1-16 v0.1.4,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/marten-seemann/qtls-go1-16/LICENSE.

   This product bundles miekg/dns v1.1.45,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/miekg/dns/LICENSE.

   This product bundles pbnjay/memory v0.0.0-20210728143218-7b4eea64cf58,
   which is available under a "BSD-3-Clause" license.
   For details, see vendor/github.com/pbnjay/memory/LICENSE.

   This product bundles gorilla/websocket v1.4.2,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/gorilla/websocket/LICENSE.

   This product bundles huin/goupnp v1.0.2,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/huin/goupnp/LICENSE.

   This product bundles pkg/errors v0.9.1,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/pkg/errors/LICENSE.

   This product bundles syndtr/goleveldb v1.0.1-0.20210819022825-2ae1ddf74ef7,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/syndtr/goleveldb/LICENSE.

   This product bundles marten-seemann/tcp v0.0.0-20210406111302-dfbc87cc63fd,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/marten-seemann/tcp/LICENSE.

   This product bundles mikioh/tcpinfo v0.0.0-20190314235526-30a79bb1804b,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/mikioh/tcpinfo/LICENSE.

   This product bundles mikioh/tcpopt v0.0.0-20190314235656-172688c1accc,
   which is available under a "BSD-2-Clause" license.
   For details, see vendor/github.com/mikioh/tcpopt/LICENSE.

```



# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

Commit:
Embed proper version by ldflags

Commit:
Set app version in goreleaser.yml

Commit:
Add our license in version command

Commit:
Add attribution for dependencies with BSD-2 or 3

lazartravica reviewed (approved):
LGTM, excellent job

Did you check that the binaries created through goreleaser were good?

lazartravica commented:
Adding a  don't merge label until we sort out the required Go version.

0xAleksaOpacic reviewed (approved):
Great job!

Kourin1996 commented:
> Did you check that the binaries created through goreleaser were good?

@lazartravica You can check the release in my repo
https://github.com/Kourin1996/polygon-sdk/releases/tag/untagged-494d5363926e266d6e6b

lazartravica commented:
> > Did you check that the binaries created through goreleaser were good?
> 
> @lazartravica You can check the release in my repo https://github.com/Kourin1996/polygon-sdk/releases/tag/untagged-494d5363926e266d6e6b

@Kourin1996  I am unable to open that link - 404  :(

Kourin1996 commented:
@lazartravica I guess you can access now
https://github.com/Kourin1996/polygon-sdk/releases/tag/v0.2.8

Commit:
Remove license and attributions from version command and add license command

Move version, license, and BSD attributions into params package

Remove blank line between license and attributions on license command

Kourin1996 commented:
I've changed few things:
* Remove license and dependency attributions from `version` command
* Add `license` command to print our license and dependency attributions for BSD
* Rename `version` package to `params` so that we can put license and attributions there
* Rename the title `DEPENDENCY LICENSES` to `DEPENDENCY ATTRIBUTIONS` in `license` command

And you can test sample of binaries/containers in the following release
https://github.com/Kourin1996/polygon-sdk/releases/tag/v0.2.13

Commit:
Fix lint error

Commit:
Merge remote-tracking branch 'origin/develop' into feature/print-version-and-license

Commit:
Add generate_dependency_licenses.sh

Commit:
Update modules dependencies to match go1.16

Signed-off-by: Luca Georges Francois <luca.georges-francois@epitech.eu>

Commit:
Update min. go version in CI tests

Signed-off-by: Luca Georges Francois <luca.georges-francois@epitech.eu>

zivkovicmilos commented on params/licenses.go:
```diff
@@ -0,0 +1,37 @@
+package params
```
What do you think about renaming this package to something more unambiguous, like `licenses` or `versioning` or something like that?

Kourin1996 replied:
It's possible, but there are not so much codes in each package as you can see

zivkovicmilos replied:
Yes, but the package name is ambiguous. Params for what? What kind of params?

I just think we can have better naming of the package, instead of `params`

Kourin1996 replied:
Fixed 9de25d8a72538a7672794d0a446e31406e8d3fac

zivkovicmilos commented on params/licenses.go:
```diff
@@ -0,0 +1,37 @@
+	License string
+
+	// Dependency Licenses
+	//go:embed bsd_licenses.json
```
Minor nitpick, space between `//` and the rest of the comment should be present

Kourin1996 replied:
This is the correct syntax. It will not be embedded if adding the space.
Please check document: https://pkg.go.dev/embed

zivkovicmilos replied:
Really weird syntax 🤦

zivkovicmilos commented on params/licenses.go:
```diff
@@ -0,0 +1,37 @@
+	// Dependency Licenses
+	//go:embed bsd_licenses.json
+	bsdLicensesJSON string
+	BsdLicenses     []DepLicense
```
These acronyms should always be upper-cased, like `BSDLicenses` and so on

Kourin1996 replied:
Renamed: 236f035318fbc226dd45dc6962a7bbc88e7b642f

zivkovicmilos commented on params/licenses.go:
```diff
@@ -0,0 +1,37 @@
+	BsdLicenses     []DepLicense
+)
+
+func init() {
+	if err := json.Unmarshal([]byte(bsdLicensesJSON), &BsdLicenses); err != nil {
+		fmt.Printf("failed to parse bsd_licenses.json")
+		os.Exit(1)
+	}
+}
```
Do we need this to be an `init` method?
Why can't we unmarshal it programmatically manually?

Kourin1996 replied:
I'm afraid I don't get what `programmatically` means. What kind of code do you image you can write?

zivkovicmilos replied:
Why do we need to rely on `init` to be present? Why can't we trigger this unmarshalling manually?

Kourin1996 replied:
I have no idea what manually means in previous comment. But let it unmarshal ad-hoc

zivkovicmilos replied:
I'll pretend we don't have this `init` method here, since we cleaned up `main` in [ac0a2e6](https://github.com/0xPolygon/polygon-edge/pull/407/commits/ac0a2e637a241829b9bda839bf60c20b1a9b0b1a) 🙂

zivkovicmilos commented on main.go:
```diff
@@ -1,13 +1,24 @@
+	license string
+)
+
+func init() {
+	params.SetLicense(license)
+}
+
```
Do we need this init?
Why can't we set this in the actual `main` method?

Kourin1996 replied:
Why do you want to set in `main` method? This is a initial flow and have `license` on the beginning, then we can call `SetLicense` in `init` function

zivkovicmilos replied:
I'm suggesting we move away from these `init` methods

Kourin1996 replied:
Sorry, I don't get it. How do you do instead of this? Putting the call in main function? I don't know what the difference is

zivkovicmilos replied:
Explicitly calling, `init` methods just hide away the code being executed

zivkovicmilos replied:
Resolving as it was fixed in [ac0a2e6](https://github.com/0xPolygon/polygon-edge/pull/407/commits/ac0a2e637a241829b9bda839bf60c20b1a9b0b1a)

zivkovicmilos reviewed (approved):
Thank you for not only adding license support, but also for adding an easy way to generate and update them in the future 🙏 

I've left a few comments that should be addressed.

I would also love to have a document on our developer docs that details how we use these tools you've provided us with, and when 💯

Commit:
Rename BsdLicsense to BSDLicense

Kourin1996 commented:
@zivkovicmilos Please note it's impossible to specify the parent path to `go:embed`

Commit:
Separate params package into licenses and versioning

Commit:
Fix main.go

Commit:
Merge branch 'develop' into feat/upgrade-minimum-go-version

Commit:
Merge commit 'refs/pull/413/head' of github.com:0xPolygon/polygon-sdk into feature/print-version-and-license

Commit:
Remove init function from main

zivkovicmilos reviewed (approved):
Thank you for resolving all of these comments 🙏 

Looks great now 💯

Commit:
Fix lint error

Commit:
Merge remote-tracking branch 'origin/develop' into feature/print-version-and-license

Commit:
Merge remote-tracking branch 'origin/feature/print-version-and-license' into feature/print-version-and-license

Commit:
Fix wrong path in Makefile

Commit:
Add archive settings in goreleaser.yml in order to ignore licenses directory to be included in zip

Commit:
Add README.md in archives in goreleaser.yml
